### PR TITLE
Honor forced illegal bench abilities during AI switch evaluation

### LINKED ISSUE #9753: Tests: AI switch evaluation ignores forced Ability(...) overrides for benched Pokemon
### Description

In battle tests, the AI's switch evaluation logic does not consistently honor test-runner forced abilities for Pokemon that are still on the bench.

This is not a normal battle-engine mechanics bug in runtime play. It is a test/simulation correctness bug: battle tests can declare a bench mon's ability with `Ability(...)`, but several AI switch paths still read the raw party ability instead of the override-aware ability that active battlers use.

The result is that AI switch logic can make the wrong decision in tests whenever the declared test ability differs from the species' stored/default ability.

One concrete example is a benched `Chandelure` forced to `ABILITY_SHADOW_TAG` in a test. During switch evaluation, the AI can still treat it as `ABILITY_FLASH_FIRE`, which prevents trapper-specific switching logic from triggering even though the test state says the benched mon has `Shadow Tag`.

The inconsistency comes from two places:

1. Several bench-evaluation paths read the party mon's ability directly with `GetMonAbility(&party[monIndex])`.
2. Switch-in simulation initializes `gBattleMons[switchinBattler]` from party data, but does not copy the override-aware ability/types path used elsewhere for test battlers.

Problematic bench reads:

```c
// src/battle_ai_switch.c
monAbility = GetMonAbility(&party[monIndex]);
```

This appears in multiple switch-evaluation helpers, including:

```c
static bool32 FindMonThatAbsorbsOpponentsMove(enum BattlerId battler)
static bool32 ShouldSwitchIfTrapperInParty(enum BattlerId battler)
static bool32 FindMonWithFlagsAndSuperEffective(enum BattlerId battler, u16 flags, u32 percentChance)
```

Problematic switch-in initialization:

```c
static void InitializeSwitchinCandidate(enum BattlerId switchinBattler, u32 monIndex, struct Pokemon *mon)
{
    u32 storeCurrBattlerPartyIndex = gBattlerPartyIndexes[switchinBattler];
    PokemonToBattleMon(mon, &gBattleMons[switchinBattler]);
    // Setup switchin battler data
    gAiThinkingStruct->saved[switchinBattler].saved = TRUE;
    SetBattlerAiData(switchinBattler, gAiLogicData);
    SetBattlerFieldStatusForSwitchin(switchinBattler);
    gBattlerPartyIndexes[switchinBattler] = monIndex;
    gAiLogicData->switchInCalc = TRUE;
    ...
}
```

By comparison, active battlers already have an override-aware path in the battle code:

```c
// src/battle_util.c
void CopyMonAbilityAndTypesToBattleMon(enum BattlerId battler, struct Pokemon *mon)
{
    gBattleMons[battler].ability = GetMonAbility(mon);
#if TESTING
    if (gTestRunnerEnabled)
    {
        u32 array = (!IsPartnerMonFromSameTrainer(battler)) ? battler : GetBattlerSide(battler);
        u32 partyIndex = gBattlerPartyIndexes[battler];
        if (TestRunner_Battle_GetForcedAbility(array, partyIndex))
            gBattleMons[battler].ability = TestRunner_Battle_GetForcedAbility(array, partyIndex);
    }
#endif
    ...
}
```

Expected behavior:

- If a battle test declares a benched mon with `Ability(ABILITY_X)`, AI switch evaluation should use that declared ability consistently.
- Bench reads and switch-in simulations should match the same test-visible battle state used for active battlers.

Observed behavior:

- Bench ability reads can ignore the forced test ability.
- AI switch decisions can therefore fail to recognize trappers, absorbers, or other ability-dependent switch candidates.

Minimal failing scenario before the fix:

- Player has `Heracross` using `Bullet Seed`.
- Opponent active mon is `Swampert`.
- Opponent benched mon is `Chandelure` forced to `Shadow Tag`.
- The AI should hard-switch into the benched trapper, but instead stays in and uses a move because the bench evaluation path does not see `Shadow Tag`.

### Reproduction Steps

1. Check out `master` before the fix, for example:

```bash
git checkout master
```

2. Add the following regression to `test/battle/ai/ai_switching.c`:

```c
AI_SINGLE_BATTLE_TEST("AI_FLAG_SMART_SWITCHING: benched forced trapper abilities are honored for hard switches")
{
    GIVEN {
        AI_FLAGS(AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_TRY_TO_FAINT | AI_FLAG_CHECK_VIABILITY | AI_FLAG_SMART_SWITCHING | AI_FLAG_SMART_MON_CHOICES | AI_FLAG_OMNISCIENT);
        PLAYER(SPECIES_HERACROSS) {
            Level(76);
            Moves(MOVE_BULLET_SEED);
            Ability(ABILITY_GUTS);
            Item(ITEM_HERACRONITE);
            Nature(NATURE_ADAMANT);
            Speed(157);
        }
        PLAYER(SPECIES_WOBBUFFET) {
            Level(1);
            Speed(1);
        }
        OPPONENT(SPECIES_SWAMPERT) {
            Level(74);
            Moves(MOVE_KNOCK_OFF, MOVE_PROTECT, MOVE_LIQUIDATION);
            Nature(NATURE_CAREFUL);
            Item(ITEM_RINDO_BERRY);
            Speed(116);
        }
        OPPONENT(SPECIES_CHANDELURE) {
            Level(74);
            Moves(MOVE_FLAMETHROWER, MOVE_ENERGY_BALL);
            Nature(NATURE_TIMID);
            Item(ITEM_CHOICE_SPECS);
            Ability(ABILITY_SHADOW_TAG);
            Speed(160);
        }
    } WHEN {
        TURN { MOVE(player, MOVE_BULLET_SEED); EXPECT_SWITCH(opponent, 1); }
    }
}
```

3. Run the targeted test:

```bash
make -j32 check TESTS="AI_FLAG_SMART_SWITCHING: benched forced trapper abilities are honored for hard switches"
```

4. Observe the failure:

```text
Expected SWITCH, got MOVE
```

5. To confirm the root cause, inspect the bench-ability reads in `src/battle_ai_switch.c`:

```c
monAbility = GetMonAbility(&party[monIndex]);
```

and compare them with the override-aware active battler ability setup in `src/battle_util.c`:

```c
CopyMonAbilityAndTypesToBattleMon(...)
```

### Images / Video

_No response_

### Version

1.15.1 (Latest release)

### Upcoming/master Version

_No response_

### Discord contact info

_No response_

### PR DESCRIPTION
**Pull request by moostoet**
## Description
This PR fixes a `TESTING` inconsistency in AI switch evaluation.

The battle test runner already supports forced illegal abilities through `Ability(...)`: if the ability is not legal for the species, the test runner stores it as a forced override and active battlers honor that override when the battle starts. Before this change, AI switch evaluation did not consistently honor that same forced state for benched mons.

As a result, a benched Pokemon could be evaluated with a different ability than the one declared by the test. That made AI switch simulations inconsistent with the active-battler state and could cause tests that rely on benched abilities to fail depending on species ability data.

The change does two things:

- `src/battle_ai_switch.c` now reads bench abilities through an override-aware helper during AI switch evaluation
- Switch-in simulation now copies ability/type data through the same override-aware path used for active battlers
- The regression has been moved out of the normal AI behavior file and into the AI test-runner coverage, because this is specifically testing the forced-illegal-ability path under `TESTING`

Affected AI switch paths include:

- `FindMonThatAbsorbsOpponentsMove`
- `ShouldSwitchIfTrapperInParty`
- `FindMonWithFlagsAndSuperEffective`
- `InitializeSwitchinCandidate`

The regression intentionally uses an illegal ability on a benched mon. That is deliberate: the goal is to exercise the test-runner override path directly so downstream species ability changes do not silently invalidate the test.

Validation:

- `make -j32 check TESTS="TESTING: forced illegal bench abilities are honored during AI switch evaluation"`
- `make -j32 check TESTS="test/battle/ai/ai_test_test_runner.c"`
- `make -j32 check TESTS="test/battle/ai/ai_switching.c"`
  - Result: `117 passed, 1 known failing`

## Media
Not applicable.

## Issue(s) that this PR fixes
Fixes #9753 

## Feature(s) this PR does NOT handle:
- Unrelated AI move-scoring bugs
- Changing normal non-test runtime behavior for how party abilities are stored
- Solving broader issue scenarios that depend on additional AI scoring or switching fixes beyond forced bench-ability handling
- Generalizing test coverage for legal bench abilities as those already resolve through normal party data

## Things to note in the release changelog:
- Fixed AI switch-evaluation test simulation so benched mons with forced illegal abilities are evaluated consistently with active battlers under `TESTING`.

## Discord contact info
viridian.

### CODE REVIEW

File: test/battle/ai/ai_switching.c [hunk 1]
```diff
@@ -2000,6 +2000,42 @@ AI_SINGLE_BATTLE_TEST("AI_FLAG_SMART_MON_CHOICES: AI considers both meeting and
...
+            Nature(NATURE_CAREFUL);
+            Item(ITEM_RINDO_BERRY);
+            Speed(116);
+        }
+        OPPONENT(SPECIES_CHANDELURE) {
+            Level(74);
+            Moves(MOVE_FLAMETHROWER, MOVE_ENERGY_BALL);
+            Nature(NATURE_TIMID);
+            Item(ITEM_CHOICE_SPECS);
+            Ability(ABILITY_SHADOW_TAG);
```
Comment (AlexOn1ine) [inline test/battle/ai/ai_switching.c]: That's technically not a legal ability which we try to avoid. 

I'm going to defer to @mrgriffin griffin though if that is a PR we want to accept.

File: test/battle/ai/ai_switching.c
Comment (hedara90) [inline test/battle/ai/ai_switching.c:1]: I think that the test should go to a test system file rather than the normal AI files, this is specifically testing the testing system.
By default Chandelure can't have Shadow Tag, so the test title should also specify that it's specifically illegal abilities that's being tested.

File: test/battle/ai/ai_switching.c (hunk 1 above)
Comment (hedara90) [reply test/battle/ai/ai_switching.c]: What we're really after here is that downstream ability changes for mons don't break tests that rely on benched mon abilities. So I think this is a valid case for illegal abilities.

File: test/battle/ai/ai_switching.c
Comment (AlexOn1ine) [reply test/battle/ai/ai_switching.c:1]: That's fine, and I agree with that. 

Just that last time we added forced abilities for tests very reluctantly so it's probably worth to get everyone here on the same page as well.

Comment (moostoet) [reply test/battle/ai/ai_switching.c:1]: @hedara90 I've implemented your changes as requested as well as reframed the PR to more accurately describe the fix that's being implemented

File: test/battle/ai/ai_switching.c (hunk 1 above)
Comment (mrgriffin) [reply test/battle/ai/ai_switching.c]: I think there's demand for this feature. I'll do a follow-up PR where (**on RHH CI only**) `Ability(...)` must be legal and there's an `ILLEGAL_ABILITY(...)` for illegal abilities. Which should hopefully draw attention to illegal abilities that could be legal during a code review.

Downstream/locally `Ability(...)` will accept anything.

Comment (AlexOn1ine) [reply test/battle/ai/ai_switching.c]: thanks for the reply!

**Review by hedara90 — APPROVED** [review]
